tools/check_wav_file_snr.m

```diff
@@ -0,0 +1,355 @@
+	% Automatic find of tone frequency
+	test_tone_f = 0;
+end
+
+if nargin < 1
+	error('Need to provide file name to analyze');
+end
+
+if exist('OCTAVE_VERSION', 'builtin')
+	pkg load signal;
+end
+
+%fp = '/home/singalsu/Downloads/';
+%wf = 'bat.wav.3Vbz0k'; % total mess
+%wf = 'bat.wav.zhJhMf'; % glitch in right channel
+%wf = 'bat.wav.Wodg37'; % low quality
+%wf = 'bat.wav.WR5of5'; % Low signal level but OK
+%wf = 'bat.wav.05IgGp'; % good
+%wf = 'bat.wav.3UiPoZ'; % good
+%fn = fullfile(fp, wf);
+
+param.t_ignore_from_start = 30e-3;
+param.t_ignore_from_end = 0;
+param.t_step = 1.0e-3;
+param.n_fft = 1024;
+param.max_snr_drop = 6;
+param.min_snr = 50;
+test_passed = true;
+do_plot = 1;
+do_print = 1;
+nfig = 0;
+
+% Read audio
+fprintf(1, 'Loading file %s\n', fn);
+[x, fs, frames, channels] = read_audio(fn);
+[~, basefn, extfn] = fileparts(fn);
+
+for ch = 1 : channels
+	% Sanity check
+	if ~signal_found(x(:, ch))
+		continue;
+	end
+
+	% STFT
+	[stft, f, t, param] = compute_stft(x(:, ch), fs, param);
+
+	% Plot
+	if do_plot
+		fnstr = sprintf('%s%s', basefn, extfn);
+		idstr = sprintf('%s ch%d', fnstr, ch);
+		nfig = nfig + 1;
+		plot_specgram(stft, f, t, idstr);
+	end
+
+	% Check frequency, get STFT frequency index
+	[signal_idx, tone_f] = find_test_tone(stft, param, f, t, test_tone_f);
+
+	% Get levels and SNR estimates
+	meas = get_tone_levels(stft, param, f, t, signal_idx);
+
+	% Plot SNR
+	if do_plot
+		plot_levels(meas, t, idstr);
+		nfig = nfig + 1;
+	end
+
+	% Checks for levels data
+	meas = check_tone_levels(stft, param, meas, t);
+
+	% If poor SNR check if there's periodic glitches
+	meas = check_glitch_periodicity(x(:,ch), param, meas, ch);
+
+	% Plot glitch waveform
+	if do_plot && meas.num_glitches > 0
+		plot_glitch(x(:, ch), param, meas, idstr);
+		nfig = nfig + 1;
+	end
+
+	all_meas(ch) = meas;
+	if meas.success == false
+		test_passed = false;
+	end
+end
+
+fprintf(1, 'Ch, Pass, Frequency, SNR min, SNR avg, Signal avg, Noise avg, Noise max, Num glitch, 1st glitch\n');
+
+for ch = 1 : channels
+	meas = all_meas(ch);
+	fprintf(1, '%d, %2d,     %5.0f,  %6.1f,  %6.1f,     %6.1f,    %6.1f,    %6.1f,     %6d,  %8.3f\n', ...
+		ch, meas.success, tone_f, meas.min_snr_db, meas.mean_snr_db, ...
+		meas.mean_signal_db, meas.mean_noise_db, ...
+		meas.max_noise_db, meas.num_glitches, meas.t_glitch);
+end
+
+if do_plot && do_print && ~test_passed
+	for i = 1 : nfig
+		pfn = sprintf('%s/analyze_%s_%d.png', logdir, fnstr, i);
+		figure(i);
+		print(pfn, '-dpng');
+	end
+end
+
+if test_passed
+	fprintf(1, 'Passed\n');
+else
+	error('Failed');
+end
+
+end
+
+%
+% Helper functions
+%
+
+function meas = check_glitch_periodicity(x, param, meas, ch)
+
+if meas.success
+	return
+end
+
+i1 = round(param.fs * param.t_ignore_from_start);
+i2 = length(x) - round(param.fs * param.t_ignore_from_end);
+[b, a] = butter(2, 0.95, 'high');
+y = abs(filter(b, a, x));
+z = y(i1:i2);
+thr = mean(z) + std(z);
+[~, locs]= findpeaks(z, 'MinPeakHeight', thr);
```

Comment:
this criteria with thr too easily claims some analog noise is periodic glitch. Should improve this to not unnecessarily report the finding. On the other hand finds quite well a periodic buffer issue.